src/fastpath/fastpath.c

```diff
@@ -6,15 +6,313 @@
+
+    /* Get the notification state */
+    uint32_t ntfnState = notification_ptr_get_state(ntfnPtr);
+
+    /* Get the notification badge */
+    word_t badge = cap_notification_cap_get_capNtfnBadge(cap);
+    switch (ntfnState) {
+    case NtfnState_Active: {
+#ifdef CONFIG_BENCHMARK_TRACK_KERNEL_ENTRIES
+        ksKernelEntry.is_fastpath = true;
+#endif
+        ntfn_set_active(ntfnPtr, badge | notification_ptr_get_ntfnMsgIdentifier(ntfnPtr));
+        restore_user_context();
+        UNREACHABLE();
+    }
+    case NtfnState_Idle: {
+        tcb_t *dest = (tcb_t *) notification_ptr_get_ntfnBoundTCB(ntfnPtr);
+
+        /* Check if we are bound and that thread is waiting for a message */
+        if (dest && thread_state_ptr_get_tsType(&dest->tcbState) == ThreadState_BlockedOnReceive) {
+            /* Basically equivalent to maybeDonateSchedContext. Check whether the thread already has
+             * a SC or if one can be donated from the notification. If neither is true, go to
+             * slowpath */
+            if (!dest->tcbSchedContext) {
+                sc = SC_PTR(notification_ptr_get_ntfnSchedContext(ntfnPtr));
+                if (sc == NULL || sc->scTcb != NULL) {
+                    slowpath(SysSend);
+                    UNREACHABLE();
+                }
+
+                /* Slowpath FPU saving case. We check that dest either isn't currently using the FPU
+                 * so that we do not have to save its state for migration (lazy FPU) or that it is the
+                 * the thread that is currently using the FPU, but the CPU is the same as that of the
+                 * scheduling context so that tcb migration is not needed. Not exactly the same as the
+                 * slowpath as this saves FPU state in the second case but there should be no reason
+                 * to touch the FPU if migration is not required */
+#ifdef ENABLE_SMP_SUPPORT
+#ifdef CONFIG_HAVE_FPU
+                if (nativeThreadUsingFPU(dest)) {
+                    slowpath(SysSend);
+                    UNREACHABLE();
+                }
+#endif
+#endif
+            } else {
+                sc = dest->tcbSchedContext;
+            }
+
+            /* Signal to higher prio thread is NOT fastpathed. Not sure if this handles idle thread */
+            if (NODE_STATE_ON_CORE(ksCurThread, sc->scCore)->tcbPriority < dest->tcbPriority) {
+                slowpath(SysSend);
+            }
+
+            /* Simplified schedContext_resume that does not change state and reverts to the
+            * slowpath in cases where the SC does not have sufficient budget, as this case
+            * adds extra scheduler logic. Normally, this is done after the sched_context donate
+            * but after tweaking it, I don't immediately see anything executed in schedContext_donate
+            * that will affect the conditions of this check */
+
+            if (sc->scRefillMax > 0 && !thread_state_get_tcbInReleaseQueue(dest->tcbState)) {
```

Comment:
I don't think that `!thread_state_get_tcbInReleaseQueue()` needs to be checked here as we know that the dest TCB is in an ep queue at this point.  (And in the other branch it is in a notification queue.

Reply:
Sorry, could you expand on this a little? I understand the point about it being in an ep queue because in the slowpath `cancel_ipc()` comes before sc donation, but I'm not 100% clear on what this means w.r.t the release queue.

Reply:
The dest thread cannot be in the release queue as it was previously either in an ep queue or a notification queue. A thread can only be in one queue at a time. So if it was already in the release queue, it wouldn't be able to be the dest thread.  In the slowpath signal, schedContext_resume() can put the dest thread into the release queue after it has removed it from the ep or notification queue, but at this point in the fastpath, that hasn't happened yet.

Reply:
In this NtfnState_Idle case, we are in the branch where `thread_state_ptr_get_tsType(&dest->tcbState) == ThreadState_BlockedOnReceive`. This is only true when the dest thread is blocked on receive on an endpoint.

Reply:
Ah right, that makes sense